sklearn/utils/estimator_checks.py

```diff
@@ -1997,3 +2083,79 @@ def check_decision_proba_consistency(name, estimator_orig):
         a = estimator.predict_proba(X_test)[:, 1]
         b = estimator.decision_function(X_test)
         assert_array_equal(rankdata(a), rankdata(b))
+
+
+def check_decision_scores_consistency(name, estimator_orig):
+    # Check that decision_function and score_samples methods of an outlier
+    # detector have outputs with perfect rank correlation.
+
+    centers = [(2, 2), (4, 4)]
+    X, _ = make_blobs(n_samples=100, random_state=0, n_features=4,
+                      centers=centers, cluster_std=1.0, shuffle=True)
+    X_test = np.random.randn(20, 2) + 4
+    estimator = clone(estimator_orig)
+    set_random_state(estimator)
+
+    estimator.fit(X)
+    a = estimator.score_samples(X_test)
+    b = estimator.decision_function(X_test)
+    assert_array_equal(rankdata(a), rankdata(b))
+
+
+def check_outliers_output_dtypes(name, estimator_orig):
+    """Check output types of outlier detectors.
+
+    Check that the predict method of outlier detectors returns int. Check also
+    that decision_function and score_samples methods return float.
+    """
+
+    # Toy sample
+    X = [[-2, -1], [-1, -1], [-1, -2], [1, 1], [1, 2], [2, 1], [5, 3], [-4, 2]]
+
+    estimator = clone(estimator_orig)
+    set_random_state(estimator)
+    estimator.fit(X)
+
+    y_pred = estimator.predict(X)
+    assert_in(y_pred.dtype, [np.dtype('int32'), np.dtype('int64')])
+    assert_array_equal(np.unique(y_pred), np.array([-1, 1]))
+
+    decision = estimator.decision_function(X)
+    assert_equal(decision.dtype, np.dtype('float'))
+
+    score = estimator.score_samples(X)
+    assert_equal(score.dtype, np.dtype('float'))
+
+
+def check_outliers_fit_predict(name, estimator_orig):
+    # Check fit_predict for outlier detectors.
+
+    X, _ = make_blobs(n_samples=300, random_state=0)
+    X = shuffle(X, random_state=7)
+    n_samples, n_features = X.shape
+    estimator = clone(estimator_orig)
+
+    set_random_state(estimator)
+
+    y_pred = estimator.fit_predict(X)
+    assert_equal(y_pred.shape, (n_samples,))
+    assert_in(y_pred.dtype, [np.dtype('int32'), np.dtype('int64')])
+    assert_array_equal(np.unique(y_pred), np.array([-1, 1]))
+
+    # check fit_predict = fit.predict when possible
```

Comment:
I don't think we should enforce this constraint necessarily. Certainly we plan to break `fit_transform(X) = fit(X).transform(X)` it for some transformers.